book-content/chapters/12-the-weird-parts.md

```diff
@@ -0,0 +1,1653 @@
+
+// ---cut---
+for (const key in yetiSeason) {
+  console.log(yetiSeason[key]);
+}
+```
+
+This is a consequence of TypeScript's open object types. TypeScript can't know the exact keys of an object at compile time, so it has to assume that there are unspecified keys on every object. The safest thing for it to do is, when you're enumerating the keys of an object, to treat them all as `string`.
+
+We'll look at a few ways to work around this in the exercises below.
+
+## The Empty Object Type
+
+Another consequence of open object types is that the empty object type `{}` doesn't behave the way you might expect.
+
+To set the stage, let's revisit the type assignability chart:
+
+![type assignability chart](https://res.cloudinary.com/total-typescript/image/upload/v1708622408/150-empty-object-type.solution.1_htrfmv.png)
+
+At the top of the chart is the `unknown` type, which can accept all other types. At the bottom is the `never` type, which no other type can be assigned to, but the `never` type itself can be assigned to any other type.
+
+Between the `never` and `unknown` types is a universe of types. The empty object type `{}` has a unique place in this universe. Instead of representing an empty object, as you might imagine, it actually represents _anything that isn't `null` or `undefined`_.
+
+This means that it can accept a number of other types: string, number, boolean, function, symbol, and objects containing properties.
+
+All of the following are valid assignments:
+
+```typescript
+const coverArtist: {} = "Guy-Manuel De Homem-Christo";
+const upcCode: {} = 724384260910;
+
+const submit = (homework: {}) => console.log(homework);
+submit("Oh Yeah");
+```
+
+However, trying to call `submit` with `null` or `undefined` will result in a TypeScript error:
+
+```ts twoslash
+// @errors: 2345
+const submit = (homework: {}) => console.log(homework);
+// ---cut---
+submit(null);
+```
+
+This might feel a bit strange. But it makes sense when you remember that TypeScript's objects are _open_. Imagine our success function actually took an object containing `message`. TypeScript would be happy if we passed it an excess property:
+
+```tsx
+const success = (response: { message: string }) =>
+  console.log(response.message);
+
+const messageWithExtra = { message: "Success!", extra: "This is extra" };
+
+success(messageWithExtra); // No Error!
+```
+
+An empty object is really the 'most open' object. Strings, numbers, booleans can all be considered objects in JavaScript. They each have properties, and methods. So TypeScript is happy to assign them to an empty object type.
+
+The only things in JavaScript that don't have properties are `null` and `undefined`. Attempting to access a property on either of these will result in a runtime error. So, they don't fit the definition of an object in TypeScript.
+
+When you consider this, the empty object type `{}` is a rather elegant solution to the problem of representing anything that isn't `null` or `undefined`.
```

Comment:
Just something else that comes to my mind that might explain what it means that TS is a structural type system:
```ts
const hasName: { name: string } = () => {} // ok, functions have `.name` property with a string type
```